File: SpectralClustering.py
from sklearn.cluster import SpectralClustering
import numpy as np
import json

def posMat(mat):
    n, m = len(mat), len(mat[0])
    for i in range(n):
        for j in range(m):
            mat[i][j] = mat[i][j] if mat[i][j] > 0 else 0

def binMat(mat):
    n, m = len(mat), len(mat[0])
    for i in range(n):
        for j in range(m):
            mat[i][j] = 1 if mat[i][j] > 0 else 0

def constructD_abs(mat):
    n, m = len(mat), len(mat[0])
    D = np.zeros((n,m))
    for i in range(n):
        D[i][i] = sum(abs(mat[i,:]))
    return D

n_clusters = 2
n_components = 2
what = 'bin'

n_samples = 10000
filename = str(n_samples) + "/sim" + str(n_samples) + ".json"
f = open(filename, 'r')
simMat = np.array(json.load(f))
print(simMat)
if what == 'pos':
    posMat(simMat)
elif what == 'bin':
    binMat(simMat)

clustering = SpectralClustering(n_clusters=n_clusters, n_components=n_components,
            affinity='precomputed', assign_labels='kmeans')
clustering.fit(simMat)

labels = {'labels': [int(label) for label in clustering.labels_]}

savefile = "labels_" + str(n_samples) + "_" + str(n_clusters) + "_" + str(n_components) + "_"  + what
f = open(savefile+".json", 'w')
json.dump(labels, f)
print(labels)

File: SpectralClustering.m
close all; clear all; clc;

n_samples = 2000
fname = num2str(n_samples) + "/sim" + num2str(n_samples) + ".json";
fid = fopen(fname, 'r');
raw = fread(fid, inf);
str = char(raw');
A = jsondecode(str);
what = 'abs'
mclc = 1

[n,m] = size(A);
D = zeros(n,m);
switch what
    case 'pos'
        for i = 1:n
            for j = 1:m
                if A(i,j) < 0
                    A(i,j) = 0;
                end
            end
        end
        for i = 1:n
            D(i,i) = sum(A(i,:));
        end
    case 'bin'
        for i = 1:n
            for j = 1:m
                if A(i,j) > 0
                    A(i,j) = 1;
                else
                    A(i,j) = 0;
                end
            end
        end
        for i = 1:n
            D(i,i) = sum(A(i,:));
        end
    case 'abs'
        for i = 1:n
            D(i,i) = sum(abs(A(i,:)));
        end
    case 'org'
        for i = 1:n
            D(i,i) = sum(A(i,:));
        end
end

Dinv = diag(1./diag(D));
D2inv = diag(1./sqrt(diag(D)));
L = eye(n) - D2inv*A*D2inv;
%L = D - A;
P = Dinv*A;

switch mclc
    case 1
        L = L + 0.1*eye(n);
        tic;
        for i = 1:5
           [vec, val] = eigs(L,2,'smallestabs');
           vec = vec(:,2);
           vec = vec/norm(vec);
           val = diag(val);
           val = val(2);
           idx = sign(vec-median(vec));
        end
        time = round(toc/5,2);
    case 2
        tic;
        for i = 1:5
           [vec, val] = eigs(P,1,'largestabs');
           vec = real(vec);
           vec = vec/norm(vec);
           idx = sign(vec - median(vec));
        end
        time = round(toc/5,2);
end

val, min(vec), max(vec)
idx(idx==-1) = 0;
savefile = "labels_" + num2str(n_samples) + "_2_2_" + what + "_" + num2str(mclc) + ".json";
fid = fopen(savefile, 'w');
labels = struct("labels", idx, "time", time, "vec", real(vec), "val", real(val));
encodedlabels = jsonencode(labels);
fprintf(fid, encodedlabels);
exit
